src/reducers/users.tsx

```diff
@@ -0,0 +1,109 @@
+import invariant from 'invariant';
+import { Reducer } from 'redux';
+import { ActionType, createAction, getType } from 'typesafe-actions';
+
+type UserId = number;
+
+export type User = {
+  id: UserId;
+  name: string;
+  email?: string;
+  permissions?: Array<string>;
+};
+
+export const actions = {
+  loadCurrentUser: createAction('LOAD_CURRENT_USER', (resolve) => {
+    return (payload: { user: User }) => resolve(payload);
+  }),
+  loadUser: createAction('LOAD_USER', (resolve) => {
+    return (payload: { user: User }) => resolve(payload);
+  }),
+  logOut: createAction('LOG_OUT', (resolve) => {
+    return () => resolve({});
+  }),
+};
+
+export type UsersState = {
+  currentUserId: UserId | null;
+  byId: { [id: number]: User };
+  byName: { [name: string]: UserId };
+};
+
+export const initialState: UsersState = {
+  currentUserId: null,
+  byId: {},
+  byName: {},
+};
+
+type LoadUserIntoStateParams = {
+  isCurrentUser?: boolean;
+  state: UsersState;
+  user: User;
+};
+
+export const loadUserIntoState = ({
```

Comment:
If you agree with my suggestion about removing the `LOAD_USER` action then you won't need this helper function and the reducer can become a lot simpler.